ccm_web/server/src/api/api.service.ts

```diff
@@ -120,6 +134,66 @@ export class APIService {
     return await sectionHandler.enrollUsers(sectionUsers)
   }
 
+  async enrollSectionExternalUsers (user: User, session: SessionData, sectionId: number, sectionUsers: SectionExternalUserDto[]): Promise<ExternalEnrollment|APIErrorData> {
+    let userRoles: CanvasRole[] = []
+    try {
+      userRoles = roleStringsToEnums(session.data.course.roles)
+    } catch (e) {
+      throw Error(`Role error for user "${user.loginId}": ${String(e.message)}`)
+    }
+
+    // FIXME: after development complete, update the following two lines
+    // const isRootAdmin: boolean = session.data.isRootAdmin // TODO: uncomment
+    const isRootAdmin = false // TODO: remove
+
+    if (isRootAdmin) {
+      logger.debug(`User "${user.loginId}" is a root admin.  ` +
+        'Skipping input role checks.')
+    } else {
+      const userAssignableRoles: string[] =
+        getRolesUserCanEnroll(userRoles).map(r => String(r))
+      if (!sectionUsers.every(sectionUser =>
+        userAssignableRoles.includes(String(sectionUser.type)))) {
+        const roleError: APIErrorData = {
+          statusCode: 403,
+          errors: [{
+            canvasStatusCode: NaN,
+            message: 'Disallowed role given. Allowed roles: ' +
+              JSON.stringify(userAssignableRoles)
+          }]
+        } as APIErrorData
+        return roleError
+      }
+    }
+
+    // Create all requested users, noting failures
+    const adminRequestor = this.canvasService.createRequestorForAdmin('/api/v1/')
+    const adminHandler = new AdminApiHandler(adminRequestor, user.loginId)
+    const canvasConfig = this.configService.get('canvas', { infer: true })
+    const createUserResponses = await adminHandler.createExternalUsers(sectionUsers, canvasConfig.newUserAccountID)
+    const newUsers = createUserResponses.filter(response => !isAPIErrorData(response)) as CanvasUserLoginEmail[]
```

Comment:
OK.  So, check each response that contains `apiErrorData` and see whether it's about the user already exists or some other error.

Here's the question:  What do we do if there are those other errors?  Certainly not go on with the next steps for the user, like enrollment.  But should they be reported as errors in some part of the response?

Reply:
This was discussed in the meeting. I think rather than `ExternalEnrollment`, the interface could be something like this:
```
interface ExternalEnrollmentsResult {
  [email: string] : {
    accountAdded: {
     success: boolean
     user?: CanvasUser
     error?: APIErrorData
   },
    invitationSent?: {
      success: boolean
      // maybe someting like error? if its available
    },
   enrollmentAdded?: {
     success: boolean
     enrollment?: CanvasEnrollment
     error?: APIErrorData
  }
}
```

You can also do clever things in TypeScript like this, using `success` as a flag.
```
interface AccountAddedSuccess {
  success: true
  user: CanvasUser
}

interface AccountAddedFailure {
  success: false
  error: APIErrorData
}

type AccountAdded = AccountAddedSuccess | AccountAddedFailure
```
Not sure that's the best way to do it, but it's an idea.

Reply:
The `success` flags may not be necessary, although they could make parsing the result simpler in the client.

Reply:
Resolving this, we've handled all this in various ways.